crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,601 @@
+                    }
+                    red_table_idx_0 &= table_idx_mask;
+                }
+                {
+                    red_table_idx_1 = expz[exp_chunk_no + 1 * (exp_digits + 1)];
+                    T = expz[exp_chunk_no + 1 + 1 * (exp_digits + 1)];
+
+                    red_table_idx_1 >>= exp_chunk_shift;
+                    /*
+                     * Get additional bits from then next quadword
+                     * when 64-bit boundaries are crossed.
+                     */
+                    if (exp_chunk_shift > 64 - exp_win_size) {
+                        T <<= (64 - exp_chunk_shift);
+                        red_table_idx_1 ^= T;
+                    }
+                    red_table_idx_1 &= table_idx_mask;
+                }
+
+                extract(&red_X[0 * red_digits], (const uint64_t*)red_table, (int)red_table_idx_0, (int)red_table_idx_1);
+            }
+
+            /* Series of squaring */
+            DAMS((uint64_t*)red_Y, (const uint64_t*)red_Y, m, k0);
+            DAMS((uint64_t*)red_Y, (const uint64_t*)red_Y, m, k0);
+            DAMS((uint64_t*)red_Y, (const uint64_t*)red_Y, m, k0);
+            DAMS((uint64_t*)red_Y, (const uint64_t*)red_Y, m, k0);
+            DAMS((uint64_t*)red_Y, (const uint64_t*)red_Y, m, k0);
+
+            damm((uint64_t*)red_Y, (const uint64_t*)red_Y, (const uint64_t*)red_X, m, k0);
+        }
+    }
+
+    /*
+     *
+     * NB: After the last AMM of exponentiation in Montgomery domain, the result
+     * may be (modlen + 1), but the conversion out of Montgomery domain
+     * performs an AMM(x,1) which guarantees that the final result is less than
+     * |m|, so no conditional subtraction is needed here. See [1] for details.
+     *
+     * [1] Gueron, S. Efficient software implementations of modular exponentiation.
+     *     DOI: 10.1007/s13389-012-0031-5
+     */
+
+    /* Convert result back in regular 2^52 domain */
+    memset(red_X, 0, 2 * red_digits * sizeof(uint64_t));
+    red_X[0 * red_digits] = 1;
+    red_X[1 * red_digits] = 1;
+    damm(out, (const uint64_t*)red_Y, (const uint64_t*)red_X, m, k0);
+
+    ret = 1;
+
+err:
+    if (storage != NULL) {
+        /* Clear whole storage */
+        OPENSSL_cleanse(storage, storage_len_bytes);
+        OPENSSL_free(storage);
+    }
+
+#undef DAMS
+    return ret;
+}
+
+// Compute the digit represented by the bytes given in |in|.
+OPENSSL_INLINE uint64_t get_digit(const uint8_t *in, int in_len)
+{
+    uint64_t digit = 0;
+
+    assert(in != NULL);
+    assert(in_len <= 8);
+
+    for (; in_len > 0; in_len--) {
+        digit <<= 8;
+        digit += (uint64_t)(in[in_len - 1]);
+    }
+    return digit;
+}
+
+/*
+ * Convert array of words in regular (base=2^64) representation to
+ * array of words in redundant (base=2^52) one. This is because the
+ * multiply/add instruction uses 52-bit representations to leave room
+ * for carries.
+ */
+static void to_words52(uint64_t *out, int out_len,
```

Comment:
I feel like this function (and `from_words52`) can be implemented in a simpler and more understandable manner. If you want to give it a try let me know, otherwise, I can try to do it myself.

Reply:
Either way! I think there may be a simpler approach to the exponent windowing as well, I'd love to follow up with a simplification pass.